### PR TITLE
Tracing instrumentation migrate to Decorator

### PR DESCRIPTION
**Pull request by tylerbenson**
Includes `@Trace`, Hystrix, and Jax-rs

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java [hunk 1]
```diff
@@ -81,6 +83,19 @@ public void put(final K key, final V value) {
       map.put(key, value);
     }
 
+    @Override
+    public void putIfAbsent(final K key, final V value) {
+      // We can't use putIfAbsent since it was added in 1.8.
+      // As a result, we must use double check locking.
+      if (!map.containsKey(key)) {
```
Comment (mar-kolya) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java]: This is potentially dangerous: key may get removed after this check leading to no key in the map after operation finishes.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/decorator/BaseDecorator.java [hunk 2]
```diff
@@ -61,4 +83,19 @@ public Span onError(final Span span, final Throwable throwable) {
     }
     return span;
   }
+
+  public String spanNameForMethod(final Method method) {
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/decorator/BaseDecorator.java]: Could you please document this.

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      for (final Path path : paths) {
+        if (path.value().startsWith("/") || (last != null && last.value().endsWith("/"))) {
+          resourceNameBuilder.append(path.value());
+        } else {
+          resourceNameBuilder.append("/");
+          resourceNameBuilder.append(path.value());
+        }
+        last = path;
+      }
+      resourceName = resourceNameBuilder.toString().trim();
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: Could you please pull out helper method that just calculates resource name and returns it - this code would be easier to read.

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  protected String spanType() {
+    return null;
+  }
+
+  @Override
+  protected String component() {
+    return "jax-rs-controller";
+  }
+
+  public void updateParent(final Scope scope, final Method method) {
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: Why is this called `updateParent`?

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    Class<?> target = method.getDeclaringClass();
+    Map<String, String> classMap = resourceNames.get(target);
+
+    if (classMap == null) {
+      resourceNames.putIfAbsent(target, new ConcurrentHashMap<String, String>());
+      classMap = resourceNames.get(target);
+    }
+
+    final String methodName = method.toString();
+    String resourceName = classMap.get(methodName);
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: Looks like you could use `Method` as a key directly.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java (hunk 1 above)
Comment (mar-kolya) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java]: At least this should be documented

Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java]: I'm not sure I follow...

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java (hunk 4 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: this jax-rs instrumentation is generally run inside an existing trace, usually started by a servlet request.  the intention is to update that parent span's resource name with the path defined by these annotations.  afterwards, a child span is created to time the actual method execution.

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java (hunk 5 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: I tried that, but it failed the caching test I wrote.  I don't think the method instances are reused between calls, and it uses the default `Object.hashcode()` implementation.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java (hunk 1 above)
Comment (mar-kolya) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/WeakMap.java]: nm, original `putIfAbsent` seems to have similar semantics.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/decorator/BaseDecorator.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/decorator/BaseDecorator.java]: Done.  Is this ok now?

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java (hunk 4 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: That 'parent span mutation' is kind of confusing, overall... but ok, thanks for explaining

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/decorator/BaseDecorator.java (hunk 2 above)
Comment (mar-kolya) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/decorator/BaseDecorator.java]: thanks!

File: dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java (hunk 5 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/jax-rs-annotations/src/main/java/datadog/trace/instrumentation/jaxrs/JaxRsAnnotationsDecorator.java]: Looks like using method as the key for the sub map as you suggested seems to work.  Using that instead.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java [hunk 6]
```diff
@@ -24,9 +26,13 @@
...
 @Slf4j
 public class HelperInjector implements Transformer {
+  // Need this because we can't put null into the injectedClassLoaders map.
+  private static final ClassLoader BOOTSTRAP_CLASSLOADER_PLACEHOLDER =
+      new SecureClassLoader(null) {};
+
   private final Set<String> helperClassNames;
   private Map<TypeDescription, byte[]> helperMap = null;
-  private final Set<ClassLoader> injectedClassLoaders = new HashSet<>();
+  private final WeakMap<ClassLoader, Boolean> injectedClassLoaders = newWeakMap();
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java]: It is actually strange that this problem has gone unnoticed for so long. Good catch!

File: dd-java-agent/instrumentation/java-concurrent/akka-2.5-testing/src/test/groovy/AkkaExecutorInstrumentationTest.groovy [hunk 7]
```diff
@@ -19,6 +20,7 @@ import java.util.concurrent.TimeUnit
  * Test executor instrumentation for Akka specific classes.
  * This is to large extent a copy of ExecutorInstrumentationTest.
  */
+@Retry
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/java-concurrent/akka-2.5-testing/src/test/groovy/AkkaExecutorInstrumentationTest.groovy]: This is terrible :(... Something very wrong is potentially going on there.

Could you please add a card to revisit this?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java (hunk 6 above)
Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java]: Indeed.  Thank goodness for good unit tests that caught this.  (Unfortunately not good enough to catch in the first place.)

File: dd-java-agent/instrumentation/java-concurrent/akka-2.5-testing/src/test/groovy/AkkaExecutorInstrumentationTest.groovy (hunk 7 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/java-concurrent/akka-2.5-testing/src/test/groovy/AkkaExecutorInstrumentationTest.groovy]: I agree.  I'm just tired of how unreliable our tests are and I don't have time to fix it properly right now.

**Pull request merged**